### PR TITLE
[RISCV] Strip W suffix from ADDIW

### PR DESCRIPTION
**Pull request by preames**
The motivation of this change is simply to reduce test duplication.  As can be seen in the (massive) test delta, we have many tests whose output differ only due to the use of addi on rv32 vs addiw on rv64 when the high bits are don't care.

As an aside, we don't need to worry about the non-zero immediate restriction on the compressed variants because we're not directly forming the compressed variants.  If we happen to get a zero immediate for the ADDI, then either a later optimization will strip the useless instruction or the encoder is responsible for not compressing the instruction.

### CODE REVIEW

**Comment by topperc** [comment]
I don't think it's safe to do this for SRAIW and SRLIW. SRAIW copies bit 31 to all the shifted right bits, SRAI copies bit 63. SRLIW inserts zeros starting at bit 31, SRLI starts at bit 63.

File: llvm/lib/Target/RISCV/RISCVOptWInstrs.cpp [hunk 1]
```diff
@@ -12,11 +12,12 @@
...
 //
-// Then it removes the -w suffix from addw, slliw and mulw instructions
-// whenever all users are dependent only on the lower word of the result of the
-// instruction. We do this only for addw, slliw, and mulw because the -w forms
-// are less compressible: c.add and c.slli have a larger register encoding than
-// their w counterparts, and there's no compressible version of mulw.
+// Then it removes the -w suffix from opw instructions whenever all users are
+// dependent only on the lower word of the result of the instruction.  This is
+// profitable for addw because c.add has a larger register encoding than c.addw.
+// For the remaining opw instructions, there is no compressed w variant. This
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVOptWInstrs.cpp:18]: There is a c.addiw.

```
C.ADDIW is an RV64C/RV128C-only instruction that performs the same computation but produces a 32-bit result, then sign-extends result to 64 bits. C.ADDIW expands into addiw rd, rd, imm. The immediate can be zero for C.ADDIW, where this corresponds to sext.w rd. C.ADDIW is only valid when rd̸=x0; the code points with rd=x0 are reserved.
```

File: llvm/lib/Target/RISCV/RISCVOptWInstrs.cpp [hunk 2]
```diff
@@ -12,11 +12,12 @@
 // extended bits aren't consumed or because the input was already sign extended
 // by an earlier instruction.
 //
-// Then it removes the -w suffix from addw, slliw and mulw instructions
-// whenever all users are dependent only on the lower word of the result of the
-// instruction. We do this only for addw, slliw, and mulw because the -w forms
-// are less compressible: c.add and c.slli have a larger register encoding than
-// their w counterparts, and there's no compressible version of mulw.
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVOptWInstrs.cpp:19]: This comment was incorrect. c.mul was added with Zcb.

**Comment by preames** [comment]
> I don't think it's safe to do this for SRAIW and SRLIW. SRAIW copies bit 31 to all the shifted right bits, SRAI copies bit 63. SRLIW inserts zeros starting at bit 31, SRLI starts at bit 63.

Are you sure about this semantic?  The wording from the specification simple says "SLLIW, SRLIW, and SRAIW are RV64I-only instructions that are analogously defined but operate on 32-bit values and sign-extend their 32-bit results to 64 bits. SLLIW, SRLIW, and SRAIW encodings with imm[5] ̸= 0 are reserved."   I wouldn't get your stated semantic from this without some very tortured reading of "operated on 32-bit values".  

Assuming you're correct about the semantic, I definitely agree these are unsound and need removed.

File: llvm/lib/Target/RISCV/RISCVOptWInstrs.cpp (hunk 1 above)
Comment (preames) [reply llvm/lib/Target/RISCV/RISCVOptWInstrs.cpp:18]: Agreed, need to fix comment.

**Comment by topperc** [comment]
> > I don't think it's safe to do this for SRAIW and SRLIW. SRAIW copies bit 31 to all the shifted right bits, SRAI copies bit 63. SRLIW inserts zeros starting at bit 31, SRLI starts at bit 63.
> 
> Are you sure about this semantic? The wording from the specification simple says "SLLIW, SRLIW, and SRAIW are RV64I-only instructions that are analogously defined but operate on 32-bit values and sign-extend their 32-bit results to 64 bits. SLLIW, SRLIW, and SRAIW encodings with imm[5] ̸= 0 are reserved." I wouldn't get your stated semantic from this without some very tortured reading of "operated on 32-bit values".
> 
> Assuming you're correct about the semantic, I definitely agree these are unsound and need removed.

I'm pretty sure. These are our isel patterns for them. Which are we get from type legalization of i32 lshr/ashr.

def : Pat<(i64 (srl (and GPR:$rs1, 0xffffffff), uimm5:$shamt)),
          (SRLIW GPR:$rs1, uimm5:$shamt)>;
def : Pat<(sra (sext_inreg GPR:$rs1, i32), uimm5:$shamt),
          (SRAIW GPR:$rs1, uimm5:$shamt)>;

**Comment by preames** [comment]
(Changes planned)

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**